pkg/daemon/metrics.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	PTPNamespace = "openshift"
+	PTPSubsystem = "ptp"
+
+	ptp4lProcessName = "ptp4l"
+	phc2sysProcessName = "phc2sys"
+)
+
+var (
+	NodeName = ""
+
+	OffsetFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: PTPNamespace,
+			Subsystem: PTPSubsystem,
+			Name: "offset_from_master",
+			Help: "",
+		},[]string{"process","node"})
+
+	StateOfPI = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: PTPNamespace,
+			Subsystem: PTPSubsystem,
+			Name: "state_of_pi",
+			Help: "",
+		},[]string{"process","node"})
+
+	FrequencyAdjustment = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: PTPNamespace,
+			Subsystem: PTPSubsystem,
+			Name: "frequency_adjustment",
+			Help: "",
+		},[]string{"process","node"})
+
+	DelayFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: PTPNamespace,
+			Subsystem: PTPSubsystem,
+			Name: "delay_from_master",
+			Help: "",
+		},[]string{"process","node"})
+)
+
+var registerMetrics sync.Once
+
+func RegisterMetrics(nodeName string) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(OffsetFromMaster)
+		prometheus.MustRegister(StateOfPI)
+		prometheus.MustRegister(FrequencyAdjustment)
+		prometheus.MustRegister(DelayFromMaster)
+
+		// Including these stats kills performance when Prometheus polls with multiple targets
+		prometheus.Unregister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+		prometheus.Unregister(prometheus.NewGoCollector())
+
+		NodeName = nodeName
+	})
+}
+
+// updatePTPMetrics ...
+func updatePTPMetrics(process string, offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster float64) {
+	OffsetFromMaster.With(prometheus.Labels{
+	"process": process,"node": NodeName}).Set(offsetFromMaster)
+
+	StateOfPI.With(prometheus.Labels{
+	"process": process,"node": NodeName}).Set(stateOfPI)
+
+	FrequencyAdjustment.With(prometheus.Labels{
+		"process": process,"node": NodeName}).Set(frequencyAdjustment)
+
+	DelayFromMaster.With(prometheus.Labels{
+		"process": process,"node": NodeName}).Set(delayFromMaster)
+}
+
+// extractMetrics ...
+func extractMetrics(processName, output string){
+	if processName == ptp4lProcessName && strings.Contains(output,"offset") {
+		offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster := extractPtp4lMetrics(output)
+		updatePTPMetrics(processName,offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster)
+	}
+
+	if processName == phc2sysProcessName && strings.Contains(output,"offset") {
+		offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster := extractPhc2sysMetrics(output)
+		updatePTPMetrics(processName,offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster)
+	}
+}
+
+func extractPtp4lMetrics(output string) (offsetFromMaster, stateOfPI, frequencyAdjustment, delayFromMaster float64) {
+	fields := strings.Fields(output)
```

Comment:
The message from ptp4l output might be slightly different when `summary_interval` directive is appended to ptp4l. We may need to detect the message format and parse accordingly.

With `summary_interval` that prints only statistics:
```
ptp4l[19779.163]: rms    6 max    9 freq  -5042 +/-   5 delay  1221 +/-   0
ptp4l[19783.163]: rms    5 max    6 freq  -5041 +/-   4 delay  1222 +/-   1
ptp4l[19787.164]: rms    8 max   16 freq  -5037 +/-   7 delay  1222 +/-   0
```
The statistics include offset root mean square (RMS), maximum absolute offset, frequency offset mean and standard deviation, and path delay mean and standard deviation. The units are nanoseconds and parts per billion (ppb). If there is only one clock update in the interval, the sample will be printed instead of the statistics.


Without `summary_interval`:
```
ptp4l[18772.973]: master offset        -24 s2 freq   -5019 path delay      1220
ptp4l[18773.973]: master offset          1 s2 freq   -5001 path delay      1218
ptp4l[18774.974]: master offset         -1 s2 freq   -5002 path delay      1218
```